pkg/utils/changenotifier.go

```diff
@@ -21,49 +21,38 @@ import (
 )
 
 type ChangeNotifier struct {
-	lock      sync.Mutex
-	observers map[string]func()
+	observers sync.Map
 }
 
 func NewChangeNotifier() *ChangeNotifier {
-	return &ChangeNotifier{
-		observers: make(map[string]func()),
-	}
+	return &ChangeNotifier{}
 }
 
 func (n *ChangeNotifier) AddObserver(key string, onChanged func()) {
-	n.lock.Lock()
-	defer n.lock.Unlock()
-
-	n.observers[key] = onChanged
+	n.observers.Store(key, onChanged)
 }
 
 func (n *ChangeNotifier) RemoveObserver(key string) {
-	n.lock.Lock()
-	defer n.lock.Unlock()
-
-	delete(n.observers, key)
+	n.observers.Delete(key)
 }
 
 func (n *ChangeNotifier) HasObservers() bool {
-	n.lock.Lock()
-	defer n.lock.Unlock()
-
-	return len(n.observers) > 0
+	hasObservers := false
+	n.observers.Range(func(key, val any) bool {
```

Comment:
@davidzhao What race caused this change? We have been removing use of `sync.Map` whenever there is a `Delete` involved due to [this](https://stackoverflow.com/questions/54174618/is-golang-sync-map-range-function-thread-safe).

Reply:
the race was in the notify code, where we have to call all of the callbacks. Thank you for the reminder about avoid sync.Map usage. I'll revert + fix it a diff way.